### PR TITLE
feat: Add Spotlight

### PR DESCRIPTION
**Pull request by HazAT**
Adds Spotlight to local Sentry

https://github.com/getsentry/sentry/assets/363802/da75151e-2b47-4b86-ab76-5cb35156be4b

### CODE REVIEW

File: static/app/bootstrap/initializeSdk.tsx [hunk 1]
```diff
@@ -169,6 +170,10 @@ export function initializeSdk(config: Config, {routes}: {routes?: Function} = {}
     },
   });
 
+  if (sentryConfig.environment === 'development') {
+    /* @__PURE__ */ Spotlight.init();
```
Comment (HazAT) [inline static/app/bootstrap/initializeSdk.tsx:174]: I don't know if the spaces (autoformatting) destroys the tree shaking here.
We don't want Spotlight in the prod code.

Comment (AbhiPrasad) [reply static/app/bootstrap/initializeSdk.tsx:174]: Isn't this like the following?

```suggestion
    /*#__PURE__*/ Spotlight.init();
```

also marking the line is fine, webpack is smart enough to recognize it.

File: webpack.config.ts [hunk 2]
```diff
@@ -612,6 +613,12 @@ if (
       },
     };
     appConfig.output!.publicPath = '/_static/dist/sentry/';
+
+    appConfig.plugins?.push(
+      new WebpackHookPlugin({
+        onBuildStart: ['node node_modules/@spotlightjs/sidecar/src/run.js'],
```
Comment (AbhiPrasad) [inline webpack.config.ts:619]: would be nice if the package should export a bin script to use

**Review by AbhiPrasad — APPROVED** [review]

File: static/app/bootstrap/initializeSdk.tsx (hunk 1 above)
Comment (HazAT) [reply static/app/bootstrap/initializeSdk.tsx]: lets see

File: webpack.config.ts (hunk 2 above)
Comment (HazAT) [reply webpack.config.ts]: We actually do, fixed

**Comment by HazAT** [comment]
I still can find spotlight in the prod build cc @AbhiPrasad :(

File: webpack.config.ts [hunk 3]
```diff
@@ -612,6 +613,12 @@ if (
       },
     };
     appConfig.output!.publicPath = '/_static/dist/sentry/';
+
+    appConfig.plugins?.push(
+      new WebpackHookPlugin({
+        onBuildStart: ['yarn run sidecar'],
```
Comment (evanpurkhiser) [inline webpack.config.ts:619]: What does this do?

Comment (HazAT) [reply webpack.config.ts:619]: Starts a node http server on 8969 until you shutdown the process

Comment (evanpurkhiser) [reply webpack.config.ts:619]: In the future would it make sense to rename this binary to `spotlight-sidecar`. Feels pretty ambiguous what this is right now.

**Comment by AbhiPrasad** [comment]
> I still can find spotlight in the prod build cc @AbhiPrasad :(

Yeah this makes sense actually because of the `sentryConfig.environment` conditional, it can't be evaluated at bundle time, so even if `sentryConfig.environment` is marked as side effect free it'll be kept.

you'll have to do

```ts
if (process.env.NODE_ENV !== 'production') {
  if (sentryConfig.environment === 'development) {
    Spotlight.init();
  }
}
```

File: yarn.lock [hunk 4]
```diff
@@ -2751,6 +2778,31 @@
...
+  integrity sha512-zBcPwAMDG2hbrPKFShSl3yukGliXTScnY3e7hu3CpvvvylAA5isc8gOINVBz3AlRlf3EquVByWk+T+yxg2QlAw==
+  dependencies:
+    "@fontsource/raleway" "^5.0.8"
+    "@sentry/utils" "^7.77.0"
+    autoprefixer "^10.4.15"
+    dayjs "^1.11.9"
+    platformicons "^5.6.2"
+    postcss "^8.4.28"
+    react "^18.2.0"
+    react-dom "^18.2.0"
```
Comment (scttcper) [inline yarn.lock:2793]: having react 18 and 17 is concerning. this should be a peer dependency with supported versions

File: yarn.lock [hunk 5]
```diff
@@ -9255,6 +9450,14 @@ platform@^1.3.3:
   resolved "https://registry.yarnpkg.com/platform/-/platform-1.3.6.tgz#48b4ce983164b209c2d45a107adb31f473a6e7a7"
   integrity sha512-fnWVljUchTro6RiCFvCXBbNhJc2NijN7oIQxbwsyL0buWJPG85v81ehlHI9fXrJsMNgTofEoWIQeClKpgxFLrg==
 
+platformicons@^5.6.2:
+  version "5.8.1"
+  resolved "https://registry.yarnpkg.com/platformicons/-/platformicons-5.8.1.tgz#b8baaba8131670a34035cd41979cfaecfdc25c39"
+  integrity sha512-/89WN1dJCqL4kK+ddblpHEQ1Z6jkLXM1PSV3zOT/dhfDDp+BhLwE1Wf72qOQ3HrHXsQTYlpybw3xrxjPUknBwg==
+  dependencies:
+    "@types/node" "*"
+    "@types/react" "*"
```
Comment (scttcper) [inline yarn.lock:9459]: double version of platformicons

**Review by scttcper — CHANGES_REQUESTED** [review]
if postcss, tailwind, etc are only used to build the package they should be moved to devDependencies and react 18 should be a peer dependency. if possible react-router-dom should be removed as well since we're not on v6

File: yarn.lock [hunk 6]
```diff
@@ -2751,6 +2778,31 @@
...
 
+"@spotlightjs/overlay@^0.0.17":
+  version "0.0.17"
+  resolved "https://registry.yarnpkg.com/@spotlightjs/overlay/-/overlay-0.0.17.tgz#d3855148937a4fa6485ac478b46f52bd88d83cf7"
+  integrity sha512-zBcPwAMDG2hbrPKFShSl3yukGliXTScnY3e7hu3CpvvvylAA5isc8gOINVBz3AlRlf3EquVByWk+T+yxg2QlAw==
+  dependencies:
+    "@fontsource/raleway" "^5.0.8"
+    "@sentry/utils" "^7.77.0"
+    autoprefixer "^10.4.15"
+    dayjs "^1.11.9"
```
Comment (scttcper) [inline yarn.lock:2789]: it doesn't really matter but this means we'll have dayjs, momentjs, and date-fns

File: yarn.lock [hunk 7]
```diff
@@ -2751,6 +2778,31 @@
...
+    platformicons "^5.6.2"
+    postcss "^8.4.28"
+    react "^18.2.0"
+    react-dom "^18.2.0"
+    react-router-dom "^6.18.0"
+    sql-formatter "^12.2.4"
+    tailwindcss "^3.3.3"
+    usehooks-ts "^2.9.1"
```
Comment (scttcper) [inline yarn.lock:2797]: not sure usehooks-ts is maintained

File: webpack.config.ts (hunk 2 above)
Comment (dcramer) [reply webpack.config.ts]: do we need to update here too @HazAT  ? https://spotlightjs.com/sidecar/webpack/

File: webpack.config.ts (hunk 3 above)
Comment (dcramer) [reply webpack.config.ts:619]: https://github.com/getsentry/spotlight/issues/165

File: yarn.lock (hunk 4 above)
Comment (dcramer) [reply yarn.lock:2793]: This shouldn't require a dep at all and we certainly wouldn't want to try to support multiple React versions. Any idea what we're doing wrong here @HazAT?

Like this isnt an ESM module right? Its fully compiled/bundled and already includes React?

**Comment by dcramer** [comment]
Before we merge lets answer why we need ANY deps to load Spotlight.

My expectation is its ENTIRELY isolated, just like the CSS is, and should have zero conflicts.

Its ok to load a 2mb local JS package...

File: webpack.config.ts (hunk 3 above)
Comment (HazAT) [reply webpack.config.ts]: It's starts the sidecar (proxy) for Spotlight

**Comment by HazAT** [comment]
@scttcper Thank you! The dependency problem has been resolved
That was critical for Spotlight

**Comment by HazAT** [comment]
> > I still can find spotlight in the prod build cc @AbhiPrasad :(
> 
> Yeah this makes sense actually because of the `sentryConfig.environment` conditional, it can't be evaluated at bundle time, so even if `sentryConfig.environment` is marked as side effect free it'll be kept.
> 
> you'll have to do
> 
> ```ts
> if (process.env.NODE_ENV !== 'production') {
>   if (sentryConfig.environment === 'development) {
>     Spotlight.init();
>   }
> }
> ```

@AbhiPrasad this worked!

File: webpack.config.ts (hunk 2 above)
Comment (HazAT) [reply webpack.config.ts]: good catch https://github.com/getsentry/spotlight/commit/cd64ee6c9cfa17291703e62f7f67e8ae405ecfd7

**Review by scttcper — APPROVED** [review]
lookin good 👍

**Review by Lms24 — APPROVED** [review]

**Comment by asottile-sentry** [comment]
reverting: broke the build

```
b"yarn run v1.22.5\n$ /workspace/node_modules/.bin/tsc -p config/tsconfig.build.json\nstatic/app/bootstrap/initializeSdk.tsx(5,28): error TS2307: Cannot find module '@spotlightjs/spotlight' or its corresponding type declarations.\nerror Command failed with exit code 2.\ninfo Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.\n"
```

**Pull request merged**